lib/ansible/modules/source_control/git.py

```diff
@@ -977,6 +1004,59 @@ def main():
     track_submodules = module.params['track_submodules']
 
     result.update(before=None)
+
+    if archive:
+        if module.check_mode:
+            result.update(changed=True)
+            module.exit_json(**result)
+
+        all_archive_fmt = {'.zip': 'zip', '.gz': 'tar.gz', '.tar': 'tar',
+                           '.tgz': 'tgz'}
+        _, archive_ext = os.path.splitext(archive)
+        archive_fmt = all_archive_fmt.get(archive_ext, None)
+        if archive_fmt is None:
+            module.fail_json(msg="Unable to get file extension from "
+                                 "archive file name : %s" % archive,
+                             details="Please specify archive as filename with "
+                                     "extension. File extension can be one "
+                                     "of ['tar', 'tar.gz', 'zip', 'tgz']")
+
+        repo_name = repo.split("/")[-1].replace(".git", "")
+        # Git archive is not supported by all git servers, so
+        # we will first clone and perform git archive from local directory
+        if dest and not os.path.exists(gitconfig):
+            clone(git_path, module, repo, dest, remote, depth, version, bare,
+                  reference, refspec, verify_commit)
+
+        if os.path.exists(archive):
+            # If git archive file exists, then create md5sum of existing file
+            # and compare it with new git archive file. if match, do nothing
+            # if does not match, then replace existing with temp archive file.
+            md5sum_old = hashlib.md5(open(archive, 'rb').read()).hexdigest()
```

Comment:
Instead of calculating the hash, you can compare the file content directly using e.g. https://docs.python.org/2/library/filecmp.html
Since you read both files full anyhow, I don't see how the hash helps.